cmd/zed/zed.c

```diff
@@ -263,18 +263,43 @@ main(int argc, char *argv[])
 	if (zed_conf_read_state(zcp, &saved_eid, saved_etime) < 0)
 		exit(EXIT_FAILURE);
 
-	zed_event_init(zcp);
+idle:
+	/*
+	 * If -I is specified, attempt to open /dev/zfs repeatedly until
+	 * successful.
+	 */
+	do {
+		if (!zed_event_init(zcp))
+			break;
+		/* Wait for some time and try again. tunable? */
+		sleep(30);
+	} while (!_got_exit && zcp->do_idle);
+
+	if (_got_exit)
+		goto out;
+
 	zed_event_seek(zcp, saved_eid, saved_etime);
 
 	while (!_got_exit) {
+		int rv;
 		if (_got_hup) {
 			_got_hup = 0;
 			(void) zed_conf_scan_dir(zcp);
 		}
-		zed_event_service(zcp);
+		rv = zed_event_service(zcp);
+
+		/* ENODEV: When kernel module is unloaded (osx) */
+		if (rv == -ENODEV)
```

Comment:
We have a somewhat related concern on Linux.  While the zed is running the open file handle prevents us from being able to unload the kernel module.  Which can be annoying, since it would be nice if we could unload/load the kmod and have the zed carry on.  So far we haven't settled on a any nice way to handle this, if you have any ideas let me know!

Oh, and `rv` here should be positive.

Reply:
Ahh so you straight up can't unload when zed runs. Interesting. 
Yeah not sure what would be a good way, SIGUSR? Sent event to tell zed to disconnect?